File: ainpc-core-plugin/src/main/java/ro/ainpc/progression/StoredProgressionSummary.java
package ro.ainpc.progression;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;

public record StoredProgressionSummary(
    int rowCount,
    int playerCount,
    int currentCount,
    int archivedCount,
    int trackedCount,
    int unresolvedDefinitionCount,
    Map<String, Integer> byStatus,
    Map<String, Integer> byTemplate,
    Map<String, Integer> byPack,
    Map<String, Integer> byMechanic,
    Map<String, Integer> byKind,
    Map<String, Integer> byCategory,
    Map<String, Integer> byScenarioKind,
    Map<String, Integer> byBaseType
) {
    public StoredProgressionSummary {
        rowCount = Math.max(0, rowCount);
        playerCount = Math.max(0, playerCount);
        currentCount = Math.max(0, currentCount);
        archivedCount = Math.max(0, archivedCount);
        trackedCount = Math.max(0, trackedCount);
        unresolvedDefinitionCount = Math.max(0, unresolvedDefinitionCount);
        byStatus = Map.copyOf(byStatus != null ? byStatus : Map.of());
        byTemplate = Map.copyOf(byTemplate != null ? byTemplate : Map.of());
        byPack = Map.copyOf(byPack != null ? byPack : Map.of());
        byMechanic = Map.copyOf(byMechanic != null ? byMechanic : Map.of());
        byKind = Map.copyOf(byKind != null ? byKind : Map.of());
        byCategory = Map.copyOf(byCategory != null ? byCategory : Map.of());
        byScenarioKind = Map.copyOf(byScenarioKind != null ? byScenarioKind : Map.of());
        byBaseType = Map.copyOf(byBaseType != null ? byBaseType : Map.of());
    }

    public static StoredProgressionSummary from(Collection<StoredProgression> progressions) {
        if (progressions == null || progressions.isEmpty()) {
            return new StoredProgressionSummary(0, 0, 0, 0, 0, 0,
                Map.of(), Map.of(), Map.of(), Map.of(), Map.of(), Map.of(), Map.of(), Map.of());
        }

        Set<String> players = new LinkedHashSet<>();
        Map<String, Integer> byStatus = new LinkedHashMap<>();
        Map<String, Integer> byTemplate = new LinkedHashMap<>();
        Map<String, Integer> byPack = new LinkedHashMap<>();
        Map<String, Integer> byMechanic = new LinkedHashMap<>();
        Map<String, Integer> byKind = new LinkedHashMap<>();
        Map<String, Integer> byCategory = new LinkedHashMap<>();
        Map<String, Integer> byScenarioKind = new LinkedHashMap<>();
        Map<String, Integer> byBaseType = new LinkedHashMap<>();
        int currentCount = 0;
        int archivedCount = 0;
        int trackedCount = 0;
        int unresolvedDefinitionCount = 0;

        for (StoredProgression progression : progressions) {
            if (progression == null) {
                continue;
            }
            if (!progression.playerUuid().isBlank()) {
                players.add(progression.playerUuid());
            }
            increment(byStatus, progression.status());
            increment(byTemplate, progression.templateId());
            increment(byPack, progression.packId());
            increment(byMechanic, progression.mechanicId());
            increment(byKind, progression.kind());
            increment(byCategory, progression.category());
            increment(byScenarioKind, progression.scenarioKind());
            increment(byBaseType, progression.baseType());
            if (progression.current()) {
                currentCount++;
            }
            if (progression.archived()) {
                archivedCount++;
            }
            if (progression.tracked()) {
                trackedCount++;
            }
            if (!progression.definitionResolved()) {
                unresolvedDefinitionCount++;
            }
        }

        return new StoredProgressionSummary(
            progressions.size(),
            players.size(),
            currentCount,
            archivedCount,
            trackedCount,
            unresolvedDefinitionCount,
            byStatus,
            byTemplate,
            byPack,
            byMechanic,
            byKind,
            byCategory,
            byScenarioKind,
            byBaseType
        );
    }

    private static void increment(Map<String, Integer> counts, String rawKey) {
        String key = rawKey == null || rawKey.isBlank() ? "unknown" : rawKey;
        counts.put(key, counts.getOrDefault(key, 0) + 1);
    }
}

File: ainpc-core-plugin/bin/main/ro/ainpc/progression/StoredProgressionSummary.kt
package ro.ainpc.progression

import java.util.LinkedHashMap
import java.util.LinkedHashSet

class StoredProgressionSummary(
    rowCount: Int,
    playerCount: Int,
    currentCount: Int,
    archivedCount: Int,
    trackedCount: Int,
    unresolvedDefinitionCount: Int,
    byStatus: Map<String, Int>?,
    byTemplate: Map<String, Int>?,
    byPack: Map<String, Int>?,
    byMechanic: Map<String, Int>?,
    byKind: Map<String, Int>?,
    byCategory: Map<String, Int>?,
    byScenarioKind: Map<String, Int>?,
    byBaseType: Map<String, Int>?
) {
    private val rowCountValue: Int = rowCount.coerceAtLeast(0)
    private val playerCountValue: Int = playerCount.coerceAtLeast(0)
    private val currentCountValue: Int = currentCount.coerceAtLeast(0)
    private val archivedCountValue: Int = archivedCount.coerceAtLeast(0)
    private val trackedCountValue: Int = trackedCount.coerceAtLeast(0)
    private val unresolvedDefinitionCountValue: Int = unresolvedDefinitionCount.coerceAtLeast(0)
    private val byStatusValue: Map<String, Int> = (byStatus ?: emptyMap()).toMap()
    private val byTemplateValue: Map<String, Int> = (byTemplate ?: emptyMap()).toMap()
    private val byPackValue: Map<String, Int> = (byPack ?: emptyMap()).toMap()
    private val byMechanicValue: Map<String, Int> = (byMechanic ?: emptyMap()).toMap()
    private val byKindValue: Map<String, Int> = (byKind ?: emptyMap()).toMap()
    private val byCategoryValue: Map<String, Int> = (byCategory ?: emptyMap()).toMap()
    private val byScenarioKindValue: Map<String, Int> = (byScenarioKind ?: emptyMap()).toMap()
    private val byBaseTypeValue: Map<String, Int> = (byBaseType ?: emptyMap()).toMap()

    fun rowCount(): Int = rowCountValue
    fun playerCount(): Int = playerCountValue
    fun currentCount(): Int = currentCountValue
    fun archivedCount(): Int = archivedCountValue
    fun trackedCount(): Int = trackedCountValue
    fun unresolvedDefinitionCount(): Int = unresolvedDefinitionCountValue
    fun byStatus(): Map<String, Int> = byStatusValue
    fun byTemplate(): Map<String, Int> = byTemplateValue
    fun byPack(): Map<String, Int> = byPackValue
    fun byMechanic(): Map<String, Int> = byMechanicValue
    fun byKind(): Map<String, Int> = byKindValue
    fun byCategory(): Map<String, Int> = byCategoryValue
    fun byScenarioKind(): Map<String, Int> = byScenarioKindValue
    fun byBaseType(): Map<String, Int> = byBaseTypeValue

    companion object {
        @JvmStatic
        fun from(progressions: Collection<StoredProgression?>?): StoredProgressionSummary {
            if (progressions.isNullOrEmpty()) {
                return StoredProgressionSummary(
                    0, 0, 0, 0, 0, 0,
                    emptyMap(), emptyMap(), emptyMap(), emptyMap(),
                    emptyMap(), emptyMap(), emptyMap(), emptyMap()
                )
            }

            val players = LinkedHashSet<String>()
            val byStatus = LinkedHashMap<String, Int>()
            val byTemplate = LinkedHashMap<String, Int>()
            val byPack = LinkedHashMap<String, Int>()
            val byMechanic = LinkedHashMap<String, Int>()
            val byKind = LinkedHashMap<String, Int>()
            val byCategory = LinkedHashMap<String, Int>()
            val byScenarioKind = LinkedHashMap<String, Int>()
            val byBaseType = LinkedHashMap<String, Int>()
            var currentCount = 0
            var archivedCount = 0
            var trackedCount = 0
            var unresolvedDefinitionCount = 0

            for (progression in progressions) {
                if (progression == null) {
                    continue
                }
                if (progression.playerUuid().isNotBlank()) {
                    players.add(progression.playerUuid())
                }
                increment(byStatus, progression.status())
                increment(byTemplate, progression.templateId())
                increment(byPack, progression.packId())
                increment(byMechanic, progression.mechanicId())
                increment(byKind, progression.kind())
                increment(byCategory, progression.category())
                increment(byScenarioKind, progression.scenarioKind())
                increment(byBaseType, progression.baseType())
                if (progression.current()) {
                    currentCount++
                }
                if (progression.archived()) {
                    archivedCount++
                }
                if (progression.tracked()) {
                    trackedCount++
                }
                if (!progression.definitionResolved()) {
                    unresolvedDefinitionCount++
                }
            }

            return StoredProgressionSummary(
                progressions.size,
                players.size,
                currentCount,
                archivedCount,
                trackedCount,
                unresolvedDefinitionCount,
                byStatus,
                byTemplate,
                byPack,
                byMechanic,
                byKind,
                byCategory,
                byScenarioKind,
                byBaseType
            )
        }

        private fun increment(counts: MutableMap<String, Int>, rawKey: String?) {
            val key = if (rawKey.isNullOrBlank()) "unknown" else rawKey
            counts[key] = (counts[key] ?: 0) + 1
        }
    }
}
